ext/soap/php_wss.c

```
@@ -0,0 +1,279 @@
+							char *buffer_content = (char *)buffer->buffer->content;
+						#endif
+						result = zend_string_init(buffer_content, strlen(buffer_content), 0);
+					}
+				}
+			}
+
+			xmlXPathFreeObject(xpath_obj);
+			xmlXPathFreeContext(xpath_ctx);
+		}
+		xmlOutputBufferClose(buffer);
+	}
+
+	return result;
+}
+/* }}} */
+
+static int add_reference_to_signed_info(zval *this_ptr, xmlNodePtr signed_info, xmlNodePtr source, zend_string *source_uri, xmlNsPtr ns_ds) /* {{{ */
+{
+	xmlNodePtr reference = xmlNewChild(signed_info, ns_ds, BAD_CAST("Reference"), NULL);
+	xmlSetProp(reference, BAD_CAST("URI"), BAD_CAST(ZSTR_VAL(source_uri)));
+	xmlNodePtr reference_trsfrms = xmlNewChild(reference, ns_ds, BAD_CAST("Transforms"), NULL);
+	xmlNodePtr reference_trsfrm = xmlNewChild(reference_trsfrms, ns_ds, BAD_CAST("Transform"), NULL);
+	xmlSetProp(reference_trsfrm, BAD_CAST("Algorithm"), BAD_CAST(SOAP_WSS_EXC_C14N));
+	xmlNodePtr reference_dm = xmlNewChild(reference, ns_ds, BAD_CAST("DigestMethod"), NULL);
+	
+	const php_hash_ops *ops = NULL;
+	zend_string *algo = NULL;
+	zend_long digest_method = Z_LVAL_P(Z_CLIENT_WSS_DIGEST_METHOD_P(this_ptr));
+	switch (digest_method) {
+		case SOAP_WSS_DIGEST_METHOD_SHA1:			
+			algo = zend_string_init("sha1", strlen("sha1"), 0);
+			ops = php_hash_fetch_ops(algo);
+			xmlSetProp(reference_dm, BAD_CAST("Algorithm"), BAD_CAST(SOAP_WSS_SHA1));
+			break;
+		case SOAP_WSS_DIGEST_METHOD_SHA256:
+			algo = zend_string_init("sha256", strlen("sha256"), 0);
+			ops = php_hash_fetch_ops(algo);
+			xmlSetProp(reference_dm, BAD_CAST("Algorithm"), BAD_CAST(SOAP_WSS_SHA256));
+			break;
+		case SOAP_WSS_DIGEST_METHOD_SHA512:
+			algo = zend_string_init("sha512", strlen("sha512"), 0);
+			ops = php_hash_fetch_ops(algo);
+			xmlSetProp(reference_dm, BAD_CAST("Algorithm"), BAD_CAST(SOAP_WSS_SHA512));			
+			break;
+	}
+
+	if (ops == NULL) {
+		if (algo != NULL) {
+			php_error_docref(NULL, E_WARNING, "invalid hashing algorithm, %s", ZSTR_VAL(algo));
```

This should throw an exception.